### ISSUE TITLE
Idea for the replacement for SARAH`Vertex function

### ISSUE THREAD

**Issue by wkotlarski** [FlexibleSUSY/FlexibleSUSY#63]
Currently inside the code that calculates a lot of observables we ask for ``SARAH`Vertex``. This is bad since it takes > 1s for every single call as the vertex has to be recalculated. Below is an idea for a replacement for this function, called `FSVertex`. It reads the cached vertices and it remembers the calls to it so that subsequent calls to the same vertex are speed up. Here's the complete code
```
FSVertexWithoutIndices[particleList_List] := FSVertexWithoutIndices[particleList] =
   Module[{allVertices,vertex, prefix = SARAH`$sarahCurrentOutputMainDir <> "/EWSB/Vertices/"},

      (* read in cached vertices *)
      allVertices = 
         Switch[Length[particleList],
            3, Get[prefix <> "VertexList3.m"],
            4, Get[prefix <> "VertexList4.m"],
            _, Print["Wrong number of particles"];Quit[1];
      ];  

      vertex = Select[allVertices, (StripFieldIndices@First[#] === particleList)&];
      Assert[Length[vertex]===1];
    
      First[vertex]
   ];  

FSVertex[particleList_List] :=
   Module[{fieldsIndices, genericFieldIndices, vertex},

      (* store indices of fields (if any) *)
      fieldsIndices = FieldIndexList /@ particleList;

      vertex = FSVertexWithoutIndices[StripFieldIndices /@ particleList];

      genericFieldIndices = FieldIndexList /@ First[vertex];

      (* restore indices *)
      vertex/.Flatten[Thread/@DeleteCases[Thread[genericFieldIndices -> fieldsIndices], _->{}]]
   ];  
```
The `FSVertex` strips field indices while the `FSVertexWithoutIndices` returns an actual vertex. Your thoughts?

**Comment by Expander** [comment]
Many thanks, this idea is great! I remember that Jobst uses something similar in his `CXXDiagrams.m` module. Maybe his implementation can be combined with your proposal?

Besides this, I have a few remarks/questions:

* Please compose the path using ``FileNameJoin[{SARAH`$sarahCurrentOutputMainDir, "EWSB", "Vertices"}]`` and `FileNameJoin[{prefix, "VertexList3.m"}]` in order to be platform independent.
* Could you please also provide unit tests to check that the `FSVertex[]` function behaves exactly as ``SARAH`Vertex[]``? I have several cases in mind that could be tested:
  - testing various particle orderings
  - testing fields with or without (or even only partially specified) indices

If `FSVertex[]` behaves exactly as ``SARAH`Vertex[]``, then it may replace the calls in `Vertices.m`, which could significantly speed up the code generation for complex models.

**Comment by SoundSonic1** [comment]
I would like to have the reading process for the cached vertices to happen only once and not every time `FSVertexWithoutIndices` is called with a different argument.

Additionally, one should check first whether the files `VertexList3.m/VertexList4.m` exist.

**Comment by wkotlarski** [comment]
@Expander, can we store the list of vertices in a "global" variable somewhere in `FlexibleSUSY`?

**Comment by wkotlarski** [comment]
The `CXXDiagrams` module from @iolojz has a private function called `LoadVerticesIfNecessary` which does what is says. After that the vertices are stored in ``SARAH`VertexList3`` and ``SARAH`VertexList4``. We might make it public. Shoud every module that uses `FSVertex` function be responsible for calling this function before calling any `FSVertex`? Also, should `LoadVerticesIfNecessary`  be moved to other module if it's not specific to `CXXDiagrams`?

**Comment by iolojz** [comment]
I like the idea of using `LoadVerticesIfNecessary[]` because it only reads the files once - afterwards the vertices are available through the SARAH variables.
However, one big issue is that vertices are only saved in a canonical ordering and one needs to manually translate the expressions to different orderings. This is also done in `VertexFunctionBodyForFields[]` in CXXDiagrams but the result is not put back into vertex form. Instead ```Vertices`SortCp[]``` is used to directly extract the relevant parts. Actually putting the result back into vertex form would require internal knowledge of ```Vertices`SortCp[]``` - i.e a few different cases. It should also be straightforward to extract the necessary code to do so. This way one could also greatly simplify `VertexFunctionBodyForFields[]` by moving the vertex fiddling into a separate function.
Such a function would need to:
- sort fields SARAH-canonically (unindexed)
- handle cases where vertices are zero
- translate the SARAH-canonically ordered expression to the desired order (```Vertices`SortCp[]```)
- take care of field indices

Most if this could be extracted `VertexFunctionBodyForFields[]`.

**Comment by iolojz** [comment]
Also, note that the FS convention for vertices is `-I` times the SARAH one.

**Comment by iolojz** [comment]
Additionally, one needs to consider GUT normalizations. SARAH does not have these, while FS does.

**Comment by Expander** [comment]
> Additionally, one needs to consider GUT normalizations. SARAH does not have these, while FS does.

Yes, I'd strongly recommend to stick to the uniform FS convention of always including the GUT normalization, everywhere. (In SARAH sometimes the GUT normalization is included, sometimes it is not, which I find very error prone.)
